### DISCUSSION TITLE
Several issues related to Data Importer

### DISCUSSION THREAD

**Discussion by MStumpp** [pimcore/pimcore#12142] [Q&A / Support]
I want to sync data between 2 systems via export/import. I use the following step-by-step approach:

1. Export on source system via CSV export:

![image001](https://user-images.githubusercontent.com/1505314/168848543-22a3e1ae-35fb-4fb1-9107-d534a6a14cba.png)

2. Import on target system via "Data Importer" (https://github.com/pimcore/data-importer) with the following resolver configuration:

![image002](https://user-images.githubusercontent.com/1505314/168848692-557e33c1-6414-48cf-919e-710b4fece8d8.png)

**Issue 1: After import, there as an additional folder level not existing on the source system:** 

Let's assume the source system has the following folder structure:
Root/Folder_1/Element_1
 
After import, an additional folder level (Folder_2) is created on the target system:
Root/Folder_1/Folder_2/Element_1

Interestingly: Name of "Folder_2" equals name of Element_1

How to get rid of Folder_2?
 
**Question 2:  How to configure the importer in order to migrate a second language into an existing object?**

Assuming the source system has fields of typ "localized fields", e.g. DE, EN

First, I exported/imported one language ("de") based on the process shown above. Everything works fine.

Configuration for one localized field:
![image003](https://user-images.githubusercontent.com/1505314/168850608-0d32b7dd-9655-467f-9194-a16e823aefbe.png)

After that, I created an export for the second language ("en") and tried to import it:
![image007](https://user-images.githubusercontent.com/1505314/168851040-1fbc3901-8487-423a-bfd2-78aaae60dc9d.png)

Unfortunately I got the following error:
"No match by NotLoadStrategy with 'Do not create' location strategy"

How to configure the importer to make this work?

**Question 3: Whats the difference between those 3 sections?**

![image008](https://user-images.githubusercontent.com/1505314/168851242-e0d1794b-bc01-4d17-9048-6603a8d29775.png)
 
**Question 4: How to import/export FieldCollections??**

Export to CSV not possible => see screenshot

![image004](https://user-images.githubusercontent.com/1505314/168851479-77151804-3b5d-4595-9729-765736f894ec.png)

Is there another way to export or import FieldCollections?

**Issue 5: Import/export of many-to-many relations crashes Pimcore**

Export looks promising:
![image006](https://user-images.githubusercontent.com/1505314/168852339-3da50055-0c05-418d-9101-ab7a908b67eb.png)

However, after import, when clicking on the data object with many-to-many relations that i just imported i get the errors similar to this:

```
Timestamp: Tue May 17 2022 17:21:16 GMT+0200 (Mitteleuropäische Sommerzeit)
Status: 500 | Internal Server Error
URL: /admin/object/save?task=publish
Method: PUT
Message: Unable to read file from location: object/g0/492/453. file_get_contents(/var/www/html/var/versions/object/g0/492/453): Failed to open stream: Permission denied
Trace:
in /var/www/html/vendor/league/flysystem/src/UnableToReadFile.php:24
#0 /var/www/html/vendor/league/flysystem/src/Local/LocalFilesystemAdapter.php(260): League\Flysystem\UnableToReadFile::fromLocation('object/g0/492/4...', 'file_get_conten...')
#1 /var/www/html/vendor/league/flysystem/src/Filesystem.php(62): League\Flysystem\Local\LocalFilesystemAdapter->read('object/g0/492/4...')
#2 /var/www/html/vendor/pimcore/pimcore/models/Version.php(369): League\Flysystem\Filesystem->read('object/g0/492/4...')
#3 /var/www/html/vendor/pimcore/pimcore/bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php(2316): Pimcore\Model\Version->loadData()
#4 /var/www/html/vendor/pimcore/pimcore/bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php(1272): Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectController->getLatestVersion(Object(Pimcore\Model\DataObject\ArtikelStandort))
#5 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(152): Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectController->saveAction(Object(Symfony\Component\HttpFoundation\Request))
#6 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(74): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#7 /var/www/html/vendor/symfony/http-kernel/Kernel.php(202): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#8 /var/www/html/public/index.php(45): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#9 {main}
```

 I checked filesystem and versions/object/g0/492/**453** does not exist.

**Comment by kingjia90** [comment]
1,2,3: 
It seems that there some misconfigurations but maybe would help with the explanation below.

Have you already watched the [Data Importer chapter](https://pimcore.com/academy/course/Data-Importer~158351/chapter-158410) in the Pimcore Academy?
In case not, i try to explain in "other" words, but you can find these info in the [documentation ](https://pimcore.com/docs/data-importer/current/Configuration/Resolver_Settings.html) as well.

With the Resolver, you define what should happen when there's a row that matches the conditions set.
The differences is that `Loading` is like a `if found/exists, replace`, `Creation` is more like `if not already existing, create` while Location Update is an `update, anything existing or is just created`.

So if you want to "sync", you likely must have Loading for existing and Creation for the new ones.

4: Fieldcollection seems not currently supported, as also stated in the screenshot. It may require some more custom implementations.
See: https://github.com/pimcore/data-importer/issues/49

5: that looks like a bug, please open a issue with more details on how to reproduce it, if it can be reproduced in the demo.

Hope it helps

**Comment by MStumpp** [comment]
@kingjia90 thanks
I was able to solve 1 to 3 via documentation and experimenting.
Regarding 5:
The error was self-caused and I got the import executing without the exception.
However, I still have problems importing more than 1x m2m relations.
I am able to successfully import a m2m relation if there is only 1x such relation.
In case of 2x or more relations, no relation is imported at all. 
Therefore, I guess the reason is the delimiter.
Question: How to correctly format a list of m2m relations in an excel sheet?

I tried the following:
- /subpath1/subpath2/,/subpath3/subpath4/
- /subpath1/subpath2/;/subpath3/subpath4/
- "/subpath1/subpath2/","/subpath3/subpath4/"
- "/subpath1/subpath2/";"/subpath3/subpath4/"
- '/subpath1/subpath2/','/subpath3/subpath4/'
- '/subpath1/subpath2/';'/subpath3/subpath4/'

None of these works.
It's also unclear how to provide metadata (= advanced m2m relation). 
As I said, it works if I only have 1x relation e.g. /path1/path2/

**Reply by kingjia90** [reply]
It doesn't matter if it's comma/semicolon separated or have single or double quote at the source, at least, it is not limited to any "predefined" rule.

You just have to set up the Transformation Pipeline (under the Mappings > *Column* ) accordigly in a way that turns "cell text" into "data"(array) using the programming functions available there. 
In your case, the `Explode` with `delimiter` should work.

**Reply by MStumpp** [reply]
Unfortunately **Explode** with **delimiter** seems not to work:
![image002](https://user-images.githubusercontent.com/1505314/170019629-9b22d86e-0768-447f-8342-3b7c257b44af.png)
Unable to save it and to move on.

Also tried with **Explode** and **Load DataObject**:
![image001](https://user-images.githubusercontent.com/1505314/170019655-a3beec11-5ddd-454e-9abf-4e3b8d5e74bf.png)
Also unable to save it and to move on.

**Reply by kingjia90** [reply]
there's a very "subtle" (i fall into this as well) *empty space* as default value before the delimiter, you have to remove it and then the load dataObject by path and you should be good to go

**Reply by MStumpp** [reply]
Thanks, explicitly removing the empty space fixed it. 

Just curious: Is it possible to import columns (in this case: "Option", "Anzahl") of an advanced many-to-many relation as well? If yes, how to format it?
![image001](https://user-images.githubusercontent.com/1505314/170050621-c5c1b769-62db-414d-9749-a74eeda00fd5.png)
![image002](https://user-images.githubusercontent.com/1505314/170050636-2e71ea76-5d1e-4fc9-8da7-ff3b3c260a75.png)

**Reply by kingjia90** [reply]
I am glad it worked! 

On the columns question, as far as i know, it would export an associative array where the keys are the column name, it may not work unless there is a feature in the pipeline that unsets the other superflous columns/keys and/or allows to pick a specific column/key. 
Could be a new feature or so, eventually i invite you to open a issue on the importer repository with more details.

So it won't work natively, unless some custom work is done during the exportation. As long as the result can be transformed into a DataObjectArray or AssetArray, then it would work.

**Reply by MStumpp** [reply]
Leaving the export aside, how should the excel file used for the import have to look like?
Currently, i have an excel file with one excel sheet column per many-to-many relation, containing a comma separated list of paths referencing data objects, e.g.
/subpath1/subpath2/,/subpath3/subpath4/
How to manually modify the excel file in order to have the columns of the advanced many-to-many relation correctly processed by the data importer? e.g. column with key:Option as shown in the screenshot

**Comment by salamunjan** [comment]
Hi, [MStumpp](https://github.com/MStumpp)

did you find solutions to problems 1,2,3? If so could you share what you did. I'm having the exact same problems.

Thank you

**Comment by BlackbitDevs** [comment]
If the goal is to fully sync 2 Pimcore systems, please have a look on https://github.com/BlackbitDigitalCommerce/pimcore-backup which has a `backup:sync` command to do exactly this. 

For partial syncs of course you need a plugin for exports and imports. We created https://pimcore.com/en/developers/marketplace/blackbit_digital_commerce/pimcore-data-director_e103850 for this. All you mentioned points are possible with this bundle. There are also some [tutorial videos](https://www.youtube.com/playlist?list=PL4-QRNfdsdKIfzQIP-c9hRruXf0r48fjt) which show your use-cases.
